File: Networking assignment arunava/C/checksum.c
#include<stdio.h>
#include<conio.h>
#include<string.h>
#include<math.h>
void check_sum(char *csum);
char data[50],cs[40];
int n,m;
void main()
{
	char s[50];
	int error,i,sum=0;
	printf("Enter the message: ");
	gets(data);
	m=strlen(data);
	printf("Enter slice length: ");
	scanf("%d",&n);
	check_sum("0");
	printf("\nCheck sum = %s",cs);
	printf("\nEnter the error position from MSB(If no error enter 0) = ");
	scanf("%d",&error);
	if(error!=0)
	{
		if(data[error-1]=='0')
			data[error-1]='1';
		else
			data[error-1]='0';
	}
	check_sum(cs);
	printf("\nCheck sum = %s",cs);
	for(i=0;i<strlen(cs);i++)
	sum+=pow(2,strlen(cs)-1-i)*(cs[i]-'0');
	if(sum==0)
	printf("\nThere is no error.");
	else
	printf("\nThere is an error.");
	getch();
}
void check_sum(char *csum)
{
	int i,extra,arr[20],j,temp,k=m,sum=0,count=0,q;
	for(i=0;i<strlen(csum);i++)
	sum+=pow(2,strlen(csum)-1-i)*(csum[i]-'0');
	extra=n-m%n;
	for(i=1;i<=extra;i++)
	data[k++]='0';
	for(i=0;i<m+extra;i+=n)
	{
		temp=0;
		for(j=i;j<i+n;j++)
		temp+=(pow(2,n-1-j%n)*(data[j]-'0'));
		sum+=temp;
	}
	for(q=sum;q!=0 && count<n;q/=2)
	cs[count++]=((q+1)%2)+'0';
	cs[count]='\0';
	strrev(cs);
}

File: Networking assignment arunava/Python/checksum.py
from math import ceil
print("Ënter the message")
mssg=input().strip()
print("Enter slice length")
sl=int(input().strip())
def check_sum_calc(mssg,cs,m):
	n=len(mssg)
	slice_num=ceil(n/m)
	slices=[]
	cs=int(cs,2)
	extra=m-n%m
	mssg+=('0'*extra)
	for i in range(0,(slice_num-1)*m+1,m):
	    slices.append(mssg[i:i+m])
	print("Slices are - ")
	arr=[]
	for item in slices:
		print(item)
		arr.append(int(item,2))
	arr.append(cs)
	check_sum=bin(sum(arr))[2:]
	if len(check_sum)>4:
		extra=len(check_sum)-m
		check_sum=check_sum[extra:]
	inv_cs=''
	for c in check_sum:
		if c=='1':
			inv_cs+='0'
		else:
			inv_cs+='1'
	return (check_sum,inv_cs)
check_sum,inv_cs=check_sum_calc(mssg,'0',sl)
print("Check sum before inversion =",check_sum)
print("Check sum after inversion =",inv_cs)
print("Transmitted message = ",mssg+inv_cs)   
check_sum_new,inv_cs_new=check_sum_calc(mssg,inv_cs,sl)
print("New check sum =",inv_cs_new,"So there is no error")
